src/Api/PubnubApi/EndPoint/PubSub/SubscribeEndpoint.cs

```diff
@@ -0,0 +1,219 @@
+        public ISubscribeOperation<T> QueryParam(Dictionary<string, object> customQueryParam)
+        {
+            this.queryParam = customQueryParam;
+            return this;
+        }
+
+        public void Execute()
+        {
+            if (this.subscribeChannelNames == null)
+            {
+                this.subscribeChannelNames = new List<string>();
+            }
+
+            if (this.subscribeChannelGroupNames == null)
+            {
+                this.subscribeChannelGroupNames = new List<string>();
+            }
+
+            if (this.presenceSubscribeEnabled)
+            {
+                List<string> presenceChannelNames = (this.subscribeChannelNames != null && this.subscribeChannelNames.Count > 0 && !string.IsNullOrEmpty(this.subscribeChannelNames[0])) 
+                                                ? this.subscribeChannelNames.Select(c => string.Format(CultureInfo.InvariantCulture, "{0}-pnpres", c)).ToList() : new List<string>();
+                List<string> presenceChannelGroupNames = (this.subscribeChannelGroupNames != null && this.subscribeChannelGroupNames.Count > 0 && !string.IsNullOrEmpty(this.subscribeChannelGroupNames[0])) 
+                                                ? this.subscribeChannelGroupNames.Select(c => string.Format(CultureInfo.InvariantCulture, "{0}-pnpres", c)).ToList() : new List<string>();
+
+                if (this.subscribeChannelNames != null && presenceChannelNames.Count > 0)
+                {
+                    this.subscribeChannelNames.AddRange(presenceChannelNames);
+                }
+
+                if (this.subscribeChannelGroupNames != null && presenceChannelGroupNames.Count > 0)
+                {
+                    this.subscribeChannelGroupNames.AddRange(presenceChannelGroupNames);
+                }
+            }
+
+            string[] channelNames = this.subscribeChannelNames != null ? this.subscribeChannelNames.ToArray() : null;
+            string[] channelGroupNames = this.subscribeChannelGroupNames != null ? this.subscribeChannelGroupNames.ToArray() : null;
+            SubscriptionCursor cursor = null;
+            if (subscribeTimetoken >= 1)
+            {
+                cursor = new SubscriptionCursor { Timetoken = subscribeTimetoken, Region = 0 };
+            }
+            Subscribe(channelNames, channelGroupNames, cursor, this.queryParam);
+        }
+
+        private void Subscribe(string[] channels, string[] channelGroups, SubscriptionCursor cursor, Dictionary<string, object> externalQueryParam)
+        {
+            if ((channels == null || channels.Length == 0) && (channelGroups == null || channelGroups.Length == 0))
+            {
+				throw new ArgumentException("Either Channel Or Channel Group or Both should be provided.");
+			}
+
+			if (this.subscribeEventEngineFactory.HasEventEngine(instanceId))
+            {
+                subscribeEventEngine = subscribeEventEngineFactory.GetEventEngine(instanceId);
+			}
+            else
```

Comment:
I don't know if it is a github issue but this formatting looks awful :kekw:

Reply:
I cross checked this.. It looks as expected in IDE

Reply:
Still Indented!

Reply:
![image](https://github.com/pubnub/c-sharp/assets/60129002/8bb66307-1212-4a5a-b08f-68e0333c2232)